### PR TITLE
DRILL-3732: Drill leaks memory if external sort hits out of disk spac…

### PR DESCRIPTION
**Pull request by adeneche**
…e exception

- ExternalSort.mergeAndSpill() cleans all it's data in case an errors occurs while it's spilling to disk
- made BatchGroup AutoCloseable so it can easily be closed with AutoCloseables.close() if an error occurs
- added injection site while External sort is spilling to disk
- added unit test that forces a 2 batch query to spill to disk and injects an exception while it does so

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java [hunk 1]
```diff
@@ -515,13 +517,23 @@ public BatchGroup mergeAndSpill(LinkedList<BatchGroup> batchGroups) throws Schem
...
+      success = true;
+    } catch (IOException e) {
+      throw UserException.resourceError(e)
+        .message("External Sort encountered an error while spilling to disk")
+        .build(logger);
+    } finally {
+      // make sure we properly release merged batch groups
       for (BatchGroup group : batchGroupList) {
-        group.cleanup();
+        AutoCloseables.close(group, logger);
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: If there is an exception during this close operation that would get ignored ...

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java [hunk 2]
```diff
@@ -506,7 +508,7 @@ public BatchGroup mergeAndSpill(LinkedList<BatchGroup> batchGroups) throws Schem
 
     String outputFile = Joiner.on("/").join(dirs.next(), fileName, spillCount++);
     BatchGroup newGroup = new BatchGroup(c1, fs, outputFile, oContext.getAllocator());
-
+    boolean threw = true;
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: Can you add comments on what is the significance of this flag ?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java [hunk 3]
```diff
@@ -514,13 +516,20 @@ public BatchGroup mergeAndSpill(LinkedList<BatchGroup> batchGroups) throws Schem
         outputContainer.setRecordCount(count);
         newGroup.addBatch(outputContainer);
       }
+      injector.injectChecked(context.getExecutionControls(), INTERRUPTION_WHILE_SPILLING, IOException.class);
       newGroup.closeOutputStream();
-      for (BatchGroup group : batchGroupList) {
-        group.cleanup();
-      }
-      hyperBatch.clear();
+      threw = false;
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: I think there is an implicit assumption here that threw = false;  statement will be the last statement in the try {} block.  Is that right ?  If so, that could be prone to some future problems if new code gets added here.

Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: `threw = false;` needs to be after any statement that could throw. I will add this in it's comment

**Comment by amansinha100** [comment]
+1 .  Rest looks good.  Pls add the comment for the 'threw = false' assignment.

**Pull request closed without merging**